sdk/core/core/src/az_url.c

```diff
@@ -0,0 +1,66 @@
+#include "az_hex_private.h"
+#include <az_precondition_internal.h>
+#include <az_url_internal.h>
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include <_az_cfg.h>
+
+AZ_NODISCARD AZ_INLINE bool _az_url_should_encode(uint8_t c)
+{
+  switch (c)
+  {
+    case '-':
+    case '_':
+    case '.':
+    case '~':
+      return false;
+    default:
+      return !(('0' <= c && c <= '9') || ('A' <= c && c <= 'Z') || ('a' <= c && c <= 'z'));
+  }
+}
+
+AZ_NODISCARD az_result _az_url_encode(az_span destination, az_span source, int32_t* out_length)
+{
+  AZ_PRECONDITION_NOT_NULL(out_length);
```

Comment:
precondition that destination size must be at least >= source?

Reply:
It should be covered by `if (az_span_size(destination) < result_size)`

Reply:
Right, I get that, but that is for the "dynamic" portion, where destination size being 2x source size might not be enough if everything needs encoding.

The precondition that destination must be at least as large as source seems like a reasonable thing the caller should make sure of. There is no case where smaller destination will succeed (since it has to be at least as big as the source).

Reply:
I think you're right. I'll add it.

Reply:
After adding it and thinking more 2 hours later, maybe it is not the best behavior, after all.
for the cases when you convert statically defined things (AZ_SPAN_FROMSTRs), it is fine. It is also fine when you have two buffers - with input and output close to each other - then you can make sure one buffer is greater than or equal to each other.
But imagine you get the string to encode from a HTTP response, as part of JSON. And you pass one of the string values from JSON to encoder. Now, you may have some idea what the reasonable size for the input data, or maybe you only have as much buffer for the encoded result, and you want to make your best effort to encode. But you don't actually control the size of the input data. Let's say it can get huge on rare occasions, and the you have no options, but to fail, and that is fine logic for your device - for example, if the response gives too long of a string, you light up a red LED and ignore the data. Now, to be fully compliant with the contract, you'd have to have an external check for both output span being >= of input, AND the return code for AZ_ERROR_INSUFFICIENT_SPAN / not being AZ_OK.

Reply:
Also, we probably shouldn't be checking for destination >= source, but `>`. `destination >= source + 2 characters`, to be exact. Because, if the size is the same, why are you calling this function at all?

Reply:
But maybe we don't need to check. I am not sure. It is a borderline decision. After all, if you have one of these case where you don't control the source size relative to destionation, so be it - write that extra size check beore invocation. In other cases, you don't have to.

Reply:
It is possible that nothing in the source requires encoding and so the destination size CAN be equal to the source size.

Reply:
> Now, to be fully compliant with the contract, you'd have to have an external check for both output span being >= of input, AND the return code for AZ_ERROR_INSUFFICIENT_SPAN / not being AZ_OK.

I understand the concern here. But I think it is worth it to avoid unnecessary O(n^2) processing in the worst case and the caller should be bailing early if it is  clear that the dest is too small. To me, the best way to approach this is to write sample usage code that highlights how callers would approach this API in the case you are concerned with. 

My intuition is, keep the precondition of (dest >= src) as it is conveys intent clearly since there is no value in calling this function otherwise,and allows bailing out early.

Reply:
FWIW, I think we should think about all this and fix it in a separate PR to let upstack consumes make progress.

For this PR, I recommend adding the precondition (or leave it as is if you prefer not to add it), add the extra tests, and check it in. :)

Reply:
@JeffreyRichter , if "It is possible that nothing in the source requires encoding and so the destination size CAN be equal to the source size" - then why to call url_encode() function at all?
If you tell me the scenario I described above - i.e. when at a compile time you can't reasonably guarantee that destination will always be greater than buffer, then why do we have "destination_size >= source_size" at all.

@ahsonkhan I think I am rather FOR the "destination_size >= source_size" after all. Exceptions, such as the one I described above, can be avoided by checking the size before invoking. But I think we need it to be "destination_size >= source_size + 2".

But I agree that "destination_size >= source_size + 2" is better to go as a separate PR.
And also I think that since we're pass things like "span_size + n" to AZ_PRECONDITION_SPAN_IS_VALID(), then SPAN_SIZE_IS_VALID() should check min_size for being >= 0, because it is possible to overflow an int, or mess up your math operations for min_sie calculations. You never want to say that "span size should be -5 or greater".

Reply:
Because the caller doesn't know whether there is any characters that require encoding. The caller has some string, they don't know what's in it. But, some of the characters MAY require encoding so this call this function to do the work. Otherwise, if the caller has to examine the string characters first, see that at least 1 needs encoding and then call the encode function, then the string could be traversed multiple times which is inefficient AND an inconvenient programming model from the caller's perspective.

The encoding will definitely always fail if the destination can't hold the source so the precondition is useful to catch this bug.

Reply:
> But I agree that "destination_size >= source_size + 2" is better to go as a separate PR.

I don't think we should do the `+ 2`.

The caller shouldn't need to know the specific of this operation (such as which specific characters are being encoded, and that encoding turns 1 char into 3 - those are impl details). 

All they care about is that sanitizing their string by calling this function will make it url safe.

Reply:
How all 3 things can be true at once?

```
void fn(az_span span_from_http_response_json)
az_span src1 = AZ_SPAN_FROM_STR("100percent") // 10 characters
az_span src2 = AZ_SPAN_FROM_STR("1hundred%") // 9 characters
az_span src3 = span_from_http_response_json; // we can't know what the size is.

uint8_t buf[10] = { 0 } // just enough to fit "100percent";
az_span_dst = AZ_SPAN_FROM_BUF(buf);

_az_url_encode(dst, src1, out); // can only succeed when a no-op, why to call encode at all?
_az_url_encode(dst, src2, out); // doomed to always fail, but preconditions will not fire
_az_url_encode(dst, src3, out); // we have no way to know, preconditions may fire depending on the input
```

Reply:
Consider a broader look which includes more real usage with appropriate buffer size and az_result checks before/after. I don't see how all three are true at once as you just showcased potential inputs and it won't be the code that folks write anyway. I think the `why to call encode at all?` question has already been answered.

Maybe we should take this offline and get to a consensus/shared understanding.

Reply:
In short, I don't understand how any of the arguments provided above is an argument against `dest_size >= source size + 2` precondition. We either don't need the size precondition, or it is +2.

Reply:
I lost some energy to find the right answer, whether the `source`, or `source+2` is correct, and a few days after I am a bit disoriented now. Maybe you are right. There are two of you at least, so I am not going to add `+2` to this precondition.